File: backend/src/routers/threads.py
from typing import Annotated

from fastapi import APIRouter, Depends, File, Form, Header, Query, Request, UploadFile

from src.bootstrap.container import get_dependency
from src.schemas.search import SimilarThreadResponse
from src.schemas.thread import ThreadCreate, ThreadDetailResponse, ThreadResponse
from src.views.dependencies import optional_bearer_token
from src.views.search_view import SearchView
from src.views.threads_view import ThreadsView

router = APIRouter(prefix="/{board_slug}/threads", tags=["threads"])


def thread_create_from_form(
    title: Annotated[str | None, Form()] = None,
    name: Annotated[str | None, Form()] = None,
    body: Annotated[str | None, Form()] = None,
    sage: Annotated[bool, Form()] = False,
) -> ThreadCreate:
    return ThreadCreate(title=title, name=name, body=body, sage=sage)


@router.get("", response_model=list[ThreadResponse])
async def list_threads(
    board_slug: str,
    request: Request,
    limit: int = 50,
    offset: int = 0,
    view: ThreadsView = Depends(lambda: get_dependency(ThreadsView)),
) -> list[ThreadResponse]:
    return await view.list_threads(board_slug, request, limit, offset)


@router.get("/similar", response_model=list[SimilarThreadResponse])
async def similar_threads_for_text(
    board_slug: str,
    request: Request,
    q: str = Query(min_length=1, max_length=200),
    view: SearchView = Depends(lambda: get_dependency(SearchView)),
) -> list[SimilarThreadResponse]:
    return await view.similar_threads_for_text(board_slug, q, request)


@router.get("/{thread_id}", response_model=ThreadDetailResponse)
async def get_thread(
    board_slug: str,
    thread_id: int,
    request: Request,
    view: ThreadsView = Depends(lambda: get_dependency(ThreadsView)),
) -> ThreadDetailResponse:
    return await view.get_thread(board_slug, thread_id, request)


@router.get("/{thread_id}/similar", response_model=list[SimilarThreadResponse])
async def similar_threads(
    board_slug: str,
    thread_id: int,
    request: Request,
    view: SearchView = Depends(lambda: get_dependency(SearchView)),
) -> list[SimilarThreadResponse]:
    return await view.similar_threads(board_slug, thread_id, request)


@router.post("", response_model=ThreadDetailResponse, status_code=201)
async def create_thread(
    board_slug: str,
    request: Request,
    data: Annotated[ThreadCreate, Depends(thread_create_from_form)],
    files: list[UploadFile] = File(default=[]),
    captcha_token: str | None = Header(alias="X-Captcha-Token", default=None),
    captcha_answer: str | None = Header(alias="X-Captcha-Answer", default=None),
    admin_token: str | None = Depends(optional_bearer_token),
    view: ThreadsView = Depends(lambda: get_dependency(ThreadsView)),
) -> ThreadDetailResponse:
    return await view.create_thread(
        board_slug, data, files, request, captcha_token, captcha_answer, admin_token
    )

File: frontend/src/api/threads.ts
import { apiClient } from '@/api/client'
import { toApiError } from '@/api/errors'
import type {
  LatestThreadResponse,
  PostResponse,
  ThreadDetailResponse,
  ThreadResponse,
} from '@/api/types'

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? ''

export async function listThreads(
  boardSlug: string,
  limit = 50,
  offset = 0,
): Promise<ThreadResponse[]> {
  const { data, error, response } = await apiClient.GET('/api/{board_slug}/threads', {
    params: { path: { board_slug: boardSlug }, query: { limit, offset } },
  })
  if (error) throw toApiError(error, response)
  return data
}

export async function listLatestThreads(limit = 5): Promise<LatestThreadResponse[]> {
  const { data, error, response } = await apiClient.GET('/api/threads/latest', {
    params: { query: { limit } },
  })
  if (error) throw toApiError(error, response)
  return data
}

export async function getThread(
  boardSlug: string,
  threadId: number,
): Promise<ThreadDetailResponse> {
  const { data, error, response } = await apiClient.GET('/api/{board_slug}/threads/{thread_id}', {
    params: { path: { board_slug: boardSlug, thread_id: threadId } },
  })
  if (error) throw toApiError(error, response)
  return data
}

export interface ThreadFields {
  title?: string
  name?: string
  body?: string
  sage?: boolean
}

// openapi-fetch does not cleanly support nested form schemas, so we use fetch directly and build FormData manually.
export async function createThread(
  boardSlug: string,
  fields: ThreadFields,
  files: File[],
  captchaToken: string,
  captchaAnswer: string,
): Promise<ThreadDetailResponse> {
  const formData = new FormData()
  if (fields.title) formData.append('title', fields.title)
  if (fields.name) formData.append('name', fields.name)
  if (fields.body) formData.append('body', fields.body)
  formData.append('sage', fields.sage ? 'true' : 'false')
  for (const file of files) {
    formData.append('files', file)
  }

  const response = await fetch(`${API_BASE}/api/${boardSlug}/threads`, {
    method: 'POST',
    headers: {
      'X-Captcha-Token': captchaToken,
      'X-Captcha-Answer': captchaAnswer,
    },
    body: formData,
  })

  if (!response.ok) {
    const body = await response.json().catch(() => ({ detail: response.statusText }))
    throw toApiError(body, response.status)
  }

  return await response.json() as Promise<ThreadDetailResponse>
}

export interface ReplyFields {
  name?: string
  body?: string
  sage?: boolean
}

// a reply needs no title and its image is optional.
export async function createReply(
  boardSlug: string,
  threadId: number,
  fields: ReplyFields,
  files: File[],
  captchaToken: string,
  captchaAnswer: string,
): Promise<PostResponse> {
  const formData = new FormData()
  if (fields.name) formData.append('name', fields.name)
  if (fields.body) formData.append('body', fields.body)
  formData.append('sage', fields.sage ? 'true' : 'false')
  for (const file of files) {
    formData.append('files', file)
  }

  const response = await fetch(`${API_BASE}/api/${boardSlug}/threads/${threadId}/posts`, {
    method: 'POST',
    headers: {
      'X-Captcha-Token': captchaToken,
      'X-Captcha-Answer': captchaAnswer,
    },
    body: formData,
  })

  if (!response.ok) {
    const body = await response.json().catch(() => ({ detail: response.statusText }))
    throw toApiError(body, response.status)
  }

  return await response.json() as Promise<PostResponse>
}

